File: src/main/java/wraith/fabricaeexnihilo/util/BlockGenerator.java
package wraith.fabricaeexnihilo.util;

import net.fabricmc.fabric.api.event.registry.RegistryEntryAddedCallback;
import net.minecraft.block.Block;
import net.minecraft.block.LeavesBlock;
import net.minecraft.item.BlockItem;
import net.minecraft.util.Identifier;
import net.minecraft.util.registry.Registry;
import wraith.fabricaeexnihilo.FabricaeExNihilo;
import wraith.fabricaeexnihilo.modules.ModBlocks;
import wraith.fabricaeexnihilo.modules.ModItems;
import wraith.fabricaeexnihilo.modules.barrels.BarrelBlock;
import wraith.fabricaeexnihilo.modules.crucibles.CrucibleBlock;
import wraith.fabricaeexnihilo.modules.infested.InfestedLeavesBlock;
import wraith.fabricaeexnihilo.modules.infested.InfestedLeavesItem;
import wraith.fabricaeexnihilo.modules.infested.NonInfestableLeavesBlock;
import wraith.fabricaeexnihilo.modules.sieves.SieveBlock;

import java.util.HashSet;

public class BlockGenerator implements wraith.fabricaeexnihilo.api.BlockGenerator {

    public static final BlockGenerator INSTANCE = new BlockGenerator();

    private BlockGenerator() {}
    
    public final HashSet<String> MOD_BLACKLIST = new HashSet<>();

    public void blacklistMod(String modid) {
        MOD_BLACKLIST.add(modid);
    }

    @Override
    public void createInfestedLeavesBlock(LeavesBlock leaves) {
        if (FabricaeExNihilo.CONFIG.modules.silkworms.enabled) {
            var originalIdentifier = Registry.BLOCK.getId(leaves);
            var infestedIdentifier = !originalIdentifier.getNamespace().equals("minecraft")
                    ? FabricaeExNihilo.id("infested_" + originalIdentifier.getNamespace() + "_" + originalIdentifier.getPath())
                    : FabricaeExNihilo.id("infested_" + originalIdentifier.getPath());
            InfestedLeavesBlock block = new InfestedLeavesBlock(leaves, ModBlocks.INFESTED_LEAVES_SETTINGS);
            Registry.register(Registry.BLOCK, infestedIdentifier, block);
            ModBlocks.INFESTED_LEAVES.put(infestedIdentifier, block);
            Registry.register(Registry.ITEM, infestedIdentifier, new InfestedLeavesItem(block, ModItems.BASE_SETTINGS));
        }
    }
    
    public void initRegistryCallBack() {
        Registry.BLOCK.forEach(block -> processBlock(Registry.BLOCK.getId(block), block));
        RegistryEntryAddedCallback.event(Registry.BLOCK).register(((index, identifier, block) -> processBlock(identifier, block)));
    }

    public void processBlock(Identifier identifier, Block block) {
        if(MOD_BLACKLIST.contains(identifier.getNamespace()) || FabricaeExNihilo.CONFIG.modules.generator.blackList.contains(identifier.toString())) {
            return;
        }
        if (block instanceof LeavesBlock leavesBlock && !(block instanceof NonInfestableLeavesBlock)) {
            createInfestedLeavesBlock(leavesBlock);
        }
    }
}

File: src/main/kotlin/exnihilofabrico/util/BlockGenerator.kt
package exnihilofabrico.util

import exnihilofabrico.ExNihiloFabrico
import exnihilofabrico.MODID
import exnihilofabrico.api.IBlockGenerator
import exnihilofabrico.id
import exnihilofabrico.modules.ModBlocks
import exnihilofabrico.modules.barrels.BarrelBlock
import exnihilofabrico.modules.crucibles.CrucibleBlock
import exnihilofabrico.modules.infested.InfestedLeavesBlock
import exnihilofabrico.modules.sieves.SieveBlock
import net.fabricmc.fabric.api.event.registry.RegistryEntryAddedCallback
import net.minecraft.block.Block
import net.minecraft.block.LeavesBlock
import net.minecraft.block.Material
import net.minecraft.item.BlockItem
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry

object BlockGenerator: IBlockGenerator {
    val MOD_BLACKLIST = mutableSetOf(MODID)

    override fun blacklistMod(modid: String) {
        MOD_BLACKLIST.add(modid)
    }

    override fun createInfestedLeavesBlock(block: LeavesBlock) {
        if(ExNihiloFabrico.config.modules.silkworms.enabled) {
            val originalIdentifier = Registry.BLOCK.getId(block)
            val infestedIdentifier =
                if(originalIdentifier.namespace != "minecraft")
                    id("infested_${originalIdentifier.namespace}_${originalIdentifier.path}")
                else
                    id("infested_${originalIdentifier.path}")
            registerBlockAndItem(InfestedLeavesBlock(block, ModBlocks.infestedLeavesSettings), infestedIdentifier)
        }
    }
    override fun createSieveBlock(plankID: Identifier, slabID: Identifier, tex: Identifier) {
        val sieveID =
            if(plankID.namespace != "minecraft")
                id("${plankID.namespace}_"+plankID.path.replace("planks","sieve"))
            else
                id(plankID.path.replace("planks","sieve"))
        registerBlockAndItem(SieveBlock(tex, plankID, slabID, ModBlocks.woodSettings), sieveID)
    }
    override fun createWoodBarrelBlock(plankID: Identifier, slabID: Identifier, tex: Identifier) {
        val barrelID =
            if(plankID.namespace != "minecraft")
                id("${plankID.namespace}_"+plankID.path.replace("planks","barrel"))
            else
                id(plankID.path.replace("planks","barrel"))
        registerBlockAndItem(BarrelBlock(tex, plankID, slabID, ModBlocks.woodSettings), barrelID)
    }
    override fun createWoodCrucibleBlock(logID: Identifier, logTex: Identifier) {
        val crucibleID =
            if(logID.namespace != "minecraft")
                id("${logID.namespace}_"+logID.path.replace("log","crucible"))
            else
                id(logID.path.replace("log","crucible"))
        registerBlockAndItem(CrucibleBlock(logTex, logID, ModBlocks.woodSettings), crucibleID)
    }

    private fun registerBlockAndItem(block: Block, identifier: Identifier) {
        Registry.register(Registry.BLOCK, identifier, block)
        Registry.register(Registry.ITEM, identifier, BlockItem(block, ModBlocks.itemSettings))
        when(block) {
            is CrucibleBlock -> ModBlocks.CRUCIBLES[identifier] = block
            is BarrelBlock -> ModBlocks.BARRELS[identifier] = block
            is SieveBlock -> ModBlocks.SIEVES[identifier] = block
            is InfestedLeavesBlock -> ModBlocks.INFESTED_LEAVES[identifier] = block
        }
    }

    fun initRegistryCallBack() {
        Registry.BLOCK.forEach {
            processBlock(Registry.BLOCK.getId(it), it)
        }
        RegistryEntryAddedCallback.event(Registry.BLOCK).register(
            RegistryEntryAddedCallback<Block>{ _, identifier, block -> processBlock(identifier, block)})
    }

    fun processBlock(identifier: Identifier, block: Block) {
        if(MOD_BLACKLIST.contains(identifier.namespace)) return
        if(ExNihiloFabrico.config.modules.generator.blackList.contains(identifier)) return
        if(block.defaultState.material == Material.WOOD && identifier.path.contains("planks")) {
            val slabID = Identifier(identifier.namespace, identifier.path.replace("planks", "slab"))
            createSieveBlock(identifier, slabID)
            createWoodBarrelBlock(identifier, slabID)
        }
        if(block.defaultState.material == Material.WOOD &&
            identifier.path.contains("log") &&
            !identifier.path.contains("stripped") &&
            !identifier.path.contains("quarter") // Suppress Terrestria's quarter blocks
        ) {
            createWoodCrucibleBlock(identifier)
        }
        else if(block is LeavesBlock) {
            createInfestedLeavesBlock(block)
        }
    }
}